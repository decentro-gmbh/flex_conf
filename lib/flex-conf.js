/**
 * @author Benjamin Assadsolimani
 * flex_conf - flexible configuration file loading.
 */

'use strict';

const debug = require('debug')('flex_conf');
const nconf = require('nconf');
const ConfigFile = require('./config-file');
const TagDefinition = require('./tag-definition');
const { getFilesFromDir } = require('./utils');

/**
 * FlexConf class, representing the entire config of the config folder
 */
class FlexConf {
  /**
   * Create a new configuration instance.
   * @param {string} configFolder - Path to the folder that holds all configuration files.
   * @param {Object} options - Configuration options.
   * @param {Object} [options.tagDefinitions={}] - Tag definitions.
   * @param {boolean} [options.loadRecursive=true] - Whether to load sub-folders of the configuration folder recursively.
   * @param {boolean} [options.folderTags=true] - Whether to parse sub-folder names as tags if applicable.
   * @param {boolean} [options.parseArgv=true] - Whether to parse command line arguments.
   * @param {boolean} [options.parseEnv=true] - Whether to parse environment variables.
   * @param {string} [options.separator='__'] - Seperator for environment variables.
   * @param {string} [options.postfix='json'] - Seperator for environment variables.
   * @param {boolean} [options.lowerCase=true] - Whether to lower-case environment variables.
   * @param {boolean} [options.autoload=true] - Whether to automatically load all configuration files on instantiation.
   */
  constructor(configFolder, options = {}) {
    // Filesystem options
    this.configFolder = configFolder;
    this.loadRecursive = options.loadRecursive !== undefined ? options.loadRecursive : true;
    this.folderTags = options.folderTags !== undefined ? options.folderTags : true;
    this.postfix = options.postfix || 'json';

    // Input sources options
    this.parseArgv = options.parseArgv !== undefined ? options.parseArgv : true;
    this.parseEnv = options.parseEnv !== undefined ? options.parseEnv : true;
    this.autoload = options.autoload !== undefined ? options.autoload : true;

    // Environment variable parsing options
    this.separator = options.separator || '__';
    this.lowerCase = options.lowerCase !== undefined ? options.lowerCase : true;

    // Instantiate tag definitions
    this.tagDefinitions = {};
    Object.keys(options.tagDefinitions || {}).forEach((tagName) => {
      this.tagDefinitions[tagName] = new TagDefinition(tagName, options.tagDefinitions[tagName]);
    });

    this.configFiles = [];

    // Create a new nconf configuration store
    this.store = new nconf.Provider();
    if (this.parseArgv) { this.store.argv(); }
    if (this.parseEnv) {
      this.store.env({
        separator: this.separator,
        lowerCase: this.lowerCase,
      });
    }

    // Load configuration files
    if (this.autoload) {
      this.loadConfigFiles();
    }
  }

  /**
   * Load all configuration files inside the root config folder and it's sub-folders if loadRecursive is activated.
   */
  loadConfigFiles() {
    // Get all config files
    const filenames = getFilesFromDir(this.configFolder, this.postfix, this.loadRecursive);
    filenames.forEach((filename) => {
      const configFile = new ConfigFile(filename, {
        tagDefinitions: this.tagDefinitions,
        configFolder: this.configFolder,
        folderTags: this.folderTags,
      });

      debug(`Processing config file: ${configFile.toString()}`);
      configFile.processFilepath(this.tagDefinitions);
      this.configFiles.push(configFile);
    });

    // Filter config files by tag values
    this.configFiles = this.configFiles.filter((configFile) => {
      return Object.keys(configFile.tags).reduce((ret, key) => {
        const value = configFile.tags[key];
        const { applies } = this.tagDefinitions[key];
        if (!applies(value)) {
          return false;
        }
        return ret && true;
      }, true);
    });

    // Sort config files by score
    this.configFiles = this.configFiles.sort((file1, file2) => file2.score - file1.score);

    this.configFiles.forEach((configFile) => {
      // Load config files in nconf store
      const json = configFile.loadJson();
      const store = {};
      store[configFile.namespace] = json;
      this.store.add(configFile.filepath, {
        type: 'literal',
        store,
      });
      debug(`Loaded configuration file ${configFile.toString()}`);
    });
  }

  /**
   * Return a final configuration object.
   */
  final() {
    return this.store.get();
  }
}

module.exports = FlexConf;
