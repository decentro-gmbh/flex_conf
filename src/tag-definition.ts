/**
 * @author Benjamin Assadsolimani
 * Tag definition class
 */

'use strict';

export class TagDefinition {
  /** Name of the tag */
  name: string;
  applies: Function;
  map: Function;
  score: Function;

  /**
   * Create a new TagDefinition instance.
   * @param name - Name of the tag.
   * @param options - Tag options.
   * @param options.applies - Function to check whether the tag value applies -> the configuration file is loaded.
   * @param options.map - Function to transform a tag value before further processing it.
   * @param options.score - Function to compute a score value for the tag.
   */
  constructor(name: string, options: {
    applies?: Function,
    map?: Function,
    score?: Function
  } = {}) {
    this.name = name;
    this.applies = options.applies || function () { return false; };
    this.map = options.map || function (val) { return val; };
    this.score = options.score || function () { return 1; };
  }
}
