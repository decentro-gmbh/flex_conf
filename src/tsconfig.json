{
  "compileOnSave": false,
  "compilerOptions": {
    "module": "commonjs",
    "target": "es2017",
    "alwaysStrict": true,
    "outDir": "../dist",
    "baseUrl": "./",
    "sourceMap": false,
    "declaration": true,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "removeComments": true,
    "typeRoots": [
      "./../node_modules/@types"
    ],
    "lib": [
      "es2017"
    ],
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "noUnusedParameters": false,
    "paths": {
    }
  }
}
