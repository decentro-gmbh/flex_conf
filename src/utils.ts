/**
 * @author Benjamin Assadsolimani
 * Utility functions of the flex_conf package
 */

'use strict';

import * as fs from 'fs';
import * as path from 'path';

/**
 * Get all <filetype> files from a directory and all its sub-directories.
 * @param {string} dirpath - Directory path (relative or absolute).
 * @param {string} filetype - Filetype that should be considered.
 * @param {boolean} recursive - Whether to get files from sub-directories.
 * @returns {Array<string>} Array of paths of the found '.<filetype>' files (sorted).
 */
export function getFilesFromDir(dirpath, filetype = null, recursive = false) {
  // Make sure path is absolute
  dirpath = path.isAbsolute(dirpath) ? dirpath : path.join(__dirname, dirpath);

  const dirpathStat = fs.statSync(dirpath);
  let files = [];

  if (dirpathStat.isDirectory()) {
    const filenames = fs.readdirSync(dirpath);
    filenames.forEach((filename) => {
      const filepath = path.join(dirpath, filename);
      const pathstat = fs.statSync(filepath);
      if (pathstat.isDirectory()) {
        // Recursively load
        if (recursive) {
          files = files.concat(getFilesFromDir(filepath, filetype, recursive));
        }
      } else if (pathstat.isFile()) {
        files.push(filepath);
      } else {
        throw new Error(`Not a directory or file: ${filepath}`);
      }
    });
  } else {
    throw new Error(`Not a directory: ${dirpath}`);
  }

  // Only consider valid <filetype> files and return them sorted
  if (filetype) {
    files = files.filter((file) => file.endsWith(`.${filetype}`));
  }
  files.sort();

  return files;
}
