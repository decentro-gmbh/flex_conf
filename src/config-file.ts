/**
 * @author Benjamin Assadsolimani
 * Configuration file
 */

'use strict';

import * as debug from 'debug';
import * as fs from 'fs';
import * as path from 'path';
import { TagDefinition } from './tag-definition';

export interface IConfigFileOptions {
  configFolder?: string;
  folderTags?: boolean;
  tagDefinitions?: {[key: string]: TagDefinition};
  tagSeparator?: string;
  keyValSeparator?: string;
}

/**
 * Configuration file class, representing a single configuration file inside the configuration folder or one of its sub-folders.
 */
export class ConfigFile {
  filepath: string;
  configFolder: string;
  folderTags: boolean;
  tagDefinitions: {[key: string]: TagDefinition};
  tagSeparator: string;
  keyValSeparator: string;
  tags: any;
  namespace: string;

  /**
   * Create a new configuration file.
   * @param filepath - Path to the configuration file.
   * @param options - Options object.
   * @param options.configFolder - Configuration files folder.
   * @param options.folderTags - Whether to parse sub-folder names as tags if applicable (default: true).
   * @param options.tagDefinitions - Tag definitions.
   * @param options.tagSeparator - Seperation character for tags inside the filename (default: '.').
   * @param options.keyValSeparator - Seperation character for a tag's key and value (default: '-').
   */
  constructor(filepath: string, configFolder: string, options: IConfigFileOptions = {}) {
    this.filepath = filepath;
    this.configFolder = configFolder;
    this.tagDefinitions = options.tagDefinitions || {};
    this.folderTags = options.folderTags;

    // Options
    this.tagSeparator = options.tagSeparator || '.';
    this.keyValSeparator = options.keyValSeparator || '-';

    // Process filepath
    this.tags = {};
    this.namespace = null;
  }

  get filename(): string { return path.basename(this.filepath); }

  get numTags(): number { return Object.keys(this.tags).length; }

  toString(): string { return `<${this.score}> ${this.filepath}`; }

  get score(): number {
    let score = 0;
    Object.keys(this.tags).forEach((key) => {
      const value = this.tags[key];
      score += this.tagDefinitions[key].score(value);
    });
    return score;
  }

  /**
   * Get an array of folder names of the config file's path. Treat the config folder as the root folder such that only folders will
   * be listed that are sub-folders of the root config folder.
   * @returns Array of folder names.
   */
  getFolders(): Array<string> {
    return path.dirname(this.filepath)
      .substr(this.configFolder.length + 1)
      .split(path.sep)
      .filter(d => d.length > 0);
  }

  /**
   * Process the path and filename of the configuration file regarding to the contained tags.
   */
  processFilepath() {
    const filetags = this.filename.split(this.tagSeparator);
    // Remove the file ending as this is not a tag
    filetags.pop();
    // The first tag is the name of the configuration file namespace
    this.namespace = filetags.shift();

    if (this.folderTags && this.configFolder) {
      // Each folder within the config directory is treated as a filetag
      const folders = this.getFolders();
      filetags.push(...folders);
    }

    filetags.forEach((filetag) => {
      const [key] = filetag.split(this.keyValSeparator);
      const value = filetag.substr(key.length + this.keyValSeparator.length);
      if (this.tagDefinitions[key]) {
        this.tags[key] = this.tagDefinitions[key].map(value);
        debug(`Extracted tag: <${key} : ${value} => ${this.tags[key]}> from filetag ${filetag}`);
      }
    });
  }

  /**
   * Load the JSON object contained by the configuration file.
   */
  loadJson(): Object {
    try {
      const data = fs.readFileSync(this.filepath, 'utf8');
      try {
        const json = JSON.parse(data);
        return json;
      } catch (err) {
        throw new Error(`Could not parse JSON of config file ${this.filepath}:\n${err}`);
      }
    } catch (err) {
      throw new Error(`Could not read config file '${this.filepath}':\n${err}`);
    }
  }
}
