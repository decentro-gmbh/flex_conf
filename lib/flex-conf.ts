/**
 * @author Benjamin Assadsolimani
 * flex_conf - flexible configuration file loading.
 */

'use strict';

import * as debug from 'debug';
import * as os from 'os';
import * as fs from 'fs';
import * as path from 'path';
import * as nconf from 'nconf';
import { ConfigFile } from './config-file'
import { TagDefinition } from './tag-definition'
import { getFilesFromDir } from './utils'

/**
 * FlexConf class, representing the entire config of the config folder
 */
export class FlexConf {
  configFolder: string;
  loadRecursive: boolean;
  folderTags: boolean;
  postfix: string;
  parseArgv: boolean;  
  parseEnv: boolean;
  autoload: boolean;
  separator: string;
  lowerCase: boolean;
  parseValues: boolean;
  tagDefinitions: Object;
  configFiles: Array<any>;
  store: any;

  /**
   * Create a new configuration instance.
   * @param {string} configFolder - Path to the folder that holds all configuration files.
   * @param {Object} options - Configuration options.
   * @param {Object} [options.tagDefinitions={}] - Tag definitions.
   * @param {boolean} [options.loadRecursive=true] - Whether to load sub-folders of the configuration folder recursively.
   * @param {boolean} [options.folderTags=true] - Whether to parse sub-folder names as tags if applicable.
   * @param {boolean} [options.parseArgv=true] - Whether to parse command line arguments.
   * @param {boolean} [options.parseEnv=true] - Whether to parse environment variables.
   * @param {string} [options.separator='__'] - Seperator for environment variables.
   * @param {string} [options.postfix='json'] - Seperator for environment variables.
   * @param {boolean} [options.lowerCase=true] - Whether to lower-case environment variables.
   * @param {boolean} [options.parseValues=false] - Attempt to parse well-known values (e.g. 'false', 'true', 'null', 'undefined', '3', '5.1' and JSON values) into their proper types. If a value cannot be parsed, it will remain a string.
   * @param {boolean} [options.autoload=true] - Whether to automatically load all configuration files on instantiation.
   */
  constructor(configFolder: string, options: any = {}) {
    // Filesystem options
    this.configFolder = configFolder;
    this.loadRecursive = options.loadRecursive !== undefined ? options.loadRecursive : true;
    this.folderTags = options.folderTags !== undefined ? options.folderTags : true;
    this.postfix = options.postfix || 'json';

    // Input sources options
    this.parseArgv = options.parseArgv !== undefined ? options.parseArgv : true;
    this.parseEnv = options.parseEnv !== undefined ? options.parseEnv : true;
    this.autoload = options.autoload !== undefined ? options.autoload : true;

    // Environment variable parsing options
    this.separator = options.separator || '__';
    this.lowerCase = options.lowerCase !== undefined ? options.lowerCase : true;
    this.parseValues = options.parseValues !== undefined ? options.parseValues : false;

    // Instantiate tag definitions
    this.tagDefinitions = {};
    Object.keys(options.tagDefinitions || {}).forEach((tagName) => {
      this.tagDefinitions[tagName] = new TagDefinition(tagName, options.tagDefinitions[tagName]);
    });

    this.configFiles = [];

    // Create a new nconf configuration store
    this.store = new nconf.Provider();
    if (this.parseArgv) { this.store.argv(); }
    if (this.parseEnv) {
      this.store.env({
        separator: this.separator,
        lowerCase: this.lowerCase,
        parseValues: this.parseValues,
      });
    }

    // Load configuration files
    if (this.autoload) {
      this.loadConfigFiles();
    }
  }

  /**
   * Load all configuration files inside the root config folder and it's sub-folders if loadRecursive is activated.
   */
  loadConfigFiles() {
    // Get all config files
    const filenames = getFilesFromDir(this.configFolder, this.postfix, this.loadRecursive);
    filenames.forEach((filename) => {
      const configFile = new ConfigFile(filename, {
        tagDefinitions: this.tagDefinitions,
        configFolder: this.configFolder,
        folderTags: this.folderTags,
      });

      debug(`Processing config file: ${configFile.toString()}`);
      configFile.processFilepath(this.tagDefinitions);
      this.configFiles.push(configFile);
    });

    // Filter config files by tag values
    this.configFiles = this.configFiles.filter((configFile) => {
      return Object.keys(configFile.tags).reduce((ret, key) => {
        const value = configFile.tags[key];
        const { applies } = this.tagDefinitions[key];
        if (!applies(value)) {
          return false;
        }
        return ret && true;
      }, true);
    });

    // Sort config files by score
    this.configFiles = this.configFiles.sort((file1, file2) => file2.score - file1.score);

    this.configFiles.forEach((configFile) => {
      // Load config files in nconf store
      const json = configFile.loadJson();
      const store = {};
      store[configFile.namespace] = json;
      this.store.add(configFile.filepath, {
        type: 'literal',
        store,
      });
      debug(`Loaded configuration file ${configFile.toString()}`);
    });
  }

  /**
   * Save a config namespace to a file.
   * @param {string} namespace - Config namespace to save to a file.
   * @param {Object} [options={}] - Options object.
   * @param {string} [options.filepath] - Path to save the config file to, defaults to "[os.tmpdir()]/[namespace].json".
   * @param {string|number} [options.space] - A String or Number object that's used to insert white space into the output JSON string for readability purposes.
   * @param {string} [options.encoding="utf8"] - File encoding, default to "utf8".
   * @param {string} [options.flag="w"] - Write operation flags, defaults to "w".
   * @param {number} [options.mode=0o600] - File permissions, default to read-only for the owner.
   * @returns {string} Path of the saved config file.
   */
  saveToFile(namespace, options: any = {}) {
    const configObject = this.store.get(namespace);
    if (!options.filepath) {
      options.filepath = path.join(os.tmpdir(), `${namespace}.json`);
    }
    fs.writeFileSync(options.filepath, JSON.stringify(configObject, null, options.space), {
      encoding: options.encoding || 'utf8',
      flag: options.flag || 'w',
    });
    fs.chmodSync(options.filepath, options.mode || 0o600);

    return options.filepath;
  }

  /**
   * Return a final configuration object.
   */
  final() {
    return this.store.get();
  }
}
