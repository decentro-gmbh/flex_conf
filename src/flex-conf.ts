/**
 * @author Benjamin Assadsolimani
 * flex_conf - flexible configuration file loading.
 */

'use strict';

import * as debug from 'debug';
import * as os from 'os';
import * as fs from 'fs';
import * as path from 'path';
import * as nconf from 'nconf';
import { ConfigFile } from './config-file';
import { TagDefinition } from './tag-definition';
import { getFilesFromDir } from './utils';

export interface IFlexConfOptions {
  /** Tag definitions */
  tagDefinitions?: Object;
  /** Whether to load sub-folders of the configuration folder recursively (default: true) */
  loadRecursive?: boolean;
  /** Whether to parse sub-folder names as tags if applicable (default: true) */
  folderTags?: boolean;
  /** Whether to parse command line arguments (default: true) */
  parseArgv?: boolean;
  /** Whether to parse environment variables (default: true) */
  parseEnv?: boolean;
  /** Seperator for environment variables (default: '__') */
  separator?: string;
  /** Seperator for environment variables (default: 'json') */
  postfix?: string;
  /** Whether to lower-case environment variables (default: false) */
  lowerCase?: boolean;
  /** Attempt to parse well-known values (e.g. 'false', 'true', 'null', 'undefined', '3', '5.1' and JSON values) into their proper types. If a value cannot be parsed, it will remain a string (default: false) */
  parseValues?: boolean;
  /** Whether to automatically load all configuration files on instantiation (default: true) */
  autoload?: boolean;
}

/**
 * FlexConf class, representing the entire config of the config folder
 */
export class FlexConf {
  public configFiles: Array<ConfigFile>;
  public store: nconf.Provider;

  private configFolder: string;
  private loadRecursive: boolean;
  private folderTags: boolean;
  private postfix: string;
  private parseArgv: boolean;
  private parseEnv: boolean;
  private autoload: boolean;
  private separator: string;
  private lowerCase: boolean;
  private parseValues: boolean;
  private tagDefinitions: { [key: string]: TagDefinition};

  /**
   * Create a new configuration instance.
   * @param configFolder - Path to the folder that holds all configuration files.
   * @param options - FlexConf configuration options.
   */
  constructor(configFolder: string, options: IFlexConfOptions = {}) {
    // Filesystem options
    this.configFolder = configFolder;
    this.loadRecursive = options.loadRecursive !== undefined ? options.loadRecursive : true;
    this.folderTags = options.folderTags !== undefined ? options.folderTags : true;
    this.postfix = options.postfix || 'json';

    // Input sources options
    this.parseArgv = options.parseArgv !== undefined ? options.parseArgv : true;
    this.parseEnv = options.parseEnv !== undefined ? options.parseEnv : true;
    this.autoload = options.autoload !== undefined ? options.autoload : true;

    // Environment variable parsing options
    this.separator = options.separator || '__';
    this.lowerCase = options.lowerCase !== undefined ? options.lowerCase : false;
    this.parseValues = options.parseValues !== undefined ? options.parseValues : false;

    // Instantiate tag definitions
    this.tagDefinitions = {};
    Object.keys(options.tagDefinitions || {}).forEach((tagName) => {
      this.tagDefinitions[tagName] = new TagDefinition(tagName, options.tagDefinitions[tagName]);
    });

    this.configFiles = [];

    // Create a new nconf configuration store
    this.store = new nconf.Provider();
    if (this.parseArgv) { this.store.argv(); }
    if (this.parseEnv) {
      this.store.env({
        separator: this.separator,
        lowerCase: this.lowerCase,
        parseValues: this.parseValues,
      });
    }

    // Load configuration files
    if (this.autoload) {
      this.loadConfigFiles();
    }
  }

  /**
   * Load all configuration files inside the root config folder and it's sub-folders if loadRecursive is activated.
   */
  loadConfigFiles(): void {
    // Get all config files
    const filenames = getFilesFromDir(this.configFolder, this.postfix, this.loadRecursive);
    filenames.forEach((filename) => {
      const configFile = new ConfigFile(filename, this.configFolder, {
        tagDefinitions: this.tagDefinitions,
        folderTags: this.folderTags,
      });

      debug(`Processing config file: ${configFile.toString()}`);
      configFile.processFilepath();
      this.configFiles.push(configFile);
    });

    // Filter config files by tag values
    this.configFiles = this.configFiles.filter((configFile) => {
      return Object.keys(configFile.tags).reduce((ret, key) => {
        const value = configFile.tags[key];
        const { applies } = this.tagDefinitions[key];
        if (!applies(value)) {
          return false;
        }
        return ret && true;
      }, true);
    });

    // Sort config files by score
    this.configFiles = this.configFiles.sort((file1, file2) => file2.score - file1.score);

    this.configFiles.forEach((configFile) => {
      // Load config files in nconf store
      const json = configFile.loadJson();
      const store = {};
      store[configFile.namespace] = json;
      this.store.add(configFile.filepath, {
        type: 'literal',
        store,
      });
      debug(`Loaded configuration file ${configFile.toString()}`);
    });
  }

  /**
   * Save a config namespace to a file.
   * @param  namespace - Config namespace to save to a file.
   * @param  options - Options object.
   * @param  options.filepath - Path to save the config file to, defaults to '[os.tmpdir()]/[namespace].json'.
   * @param  options.space - A String or Number object that's used to insert white space into the output JSON string for readability purposes.
   * @param  options.encoding - File encoding, default to 'utf8'.
   * @param  options.flag - Write operation flags, defaults to 'w'.
   * @param  options.mode - File permissions, default to read-only for the owner (0o600).
   * @returns Path of the saved config file.
   */
  saveToFile(namespace: string, options: {
    filepath?: string,
    space?: string|number,
    encoding?: string
    flag?: string
    mode?: number,
  } = {}): string {
    const configObject = this.store.get(namespace);
    if (!options.filepath) {
      options.filepath = path.join(os.tmpdir(), `${namespace}.json`);
    }
    fs.writeFileSync(options.filepath, JSON.stringify(configObject, null, options.space), {
      encoding: options.encoding || 'utf8',
      flag: options.flag || 'w',
    });
    fs.chmodSync(options.filepath, options.mode || 0o600);

    return options.filepath;
  }

  /**
   * Return a final configuration object.
   */
  final(): any {
    return this.store.get();
  }
}
