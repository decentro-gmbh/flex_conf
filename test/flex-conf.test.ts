/**
 *
 */

 import { FlexConf } from '../src/flex-conf';
 import { join } from 'path';

 describe('Correct loading of a specific configuration file', () => {
   test('Tag-based configuration file loading', () => {
     for (let tag1 = 1; tag1 <= 2; tag1 += 1) {
      for (let tag2 = 1; tag2 <= 2; tag2 += 1) {
        const flexConf = new FlexConf(join(__dirname, 'configs'), {
          folderTags: true,
          tagDefinitions: {
            tag1: {
             applies: val => val === `val${tag1}`
            },
            tag2: {
             applies: val => val === `val${tag2}`
            }
          }
         });
         const config = flexConf.final();
         expect(config.name1).toMatchSnapshot();
      }
     }

   });
 });
