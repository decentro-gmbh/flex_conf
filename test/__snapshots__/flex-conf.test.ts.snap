// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correct loading of a specific configuration file Tag-based configuration file loading 1`] = `
Object {
  "attr1": "tag1-val1.tag2-val1.attr1",
  "attr2": "tag1-val1.tag2-val1.attr2",
  "obj1": Object {
    "attr3": "tag1-val1.tag2-val1.attr3",
  },
}
`;

exports[`Correct loading of a specific configuration file Tag-based configuration file loading 2`] = `
Object {
  "attr1": "tag1-val1.tag2-val2.attr1",
  "attr2": "tag1-val1.tag2-val2.attr2",
  "obj1": Object {
    "attr3": "tag1-val1.tag2-val2.attr3",
  },
}
`;

exports[`Correct loading of a specific configuration file Tag-based configuration file loading 3`] = `
Object {
  "attr1": "tag2-val1.tag1-val2.attr1",
  "attr2": "tag2-val1.tag1-val2.attr2",
  "obj1": Object {
    "attr3": "tag2-val1.tag1-val2.attr3",
  },
}
`;

exports[`Correct loading of a specific configuration file Tag-based configuration file loading 4`] = `
Object {
  "attr1": "tag1-val2.tag2-val2.attr1",
  "attr2": "tag1-val2.tag2-val2.attr2",
  "obj1": Object {
    "attr3": "tag1-val2.tag2-val2.attr3",
  },
}
`;
